Feature: Two-line entries for variable-name dropdowns

arikorn opened the pull request:
This PR makes the variable-name dropdown menus behave in actions and feedbacks similarly to the way they work in the expression editor. (_**Edit: I updated the formatting below -- see the second comment, also edited. Although this first approach may have advantages in simplicity the other is more consistent with the existing UX**_...)

Each entry in the dropdown is two lines:
~~~
$(module:varname)
Variable description text
~~~

<img width="609" height="234" alt="image" src="https://github.com/user-attachments/assets/57370796-a68f-4fac-a88b-708a867e8780" />

<BR/>
<BR/>

NB
----
The main effect of the CSS change in this PR is to allow `\n` to break lines in menu-items. In addition, it removes some "excessive" spacing between menu items.

Wrapping behavior is not changed: Prior to this PR long values were wrapped in menu items but not in the current value (i.e. when dropdown is not "open). This PR does not change the behavior, but it would be very easy to wrap text even in the current value -- in fact I had to do extra work to prevent it!

Conversely, if preferred: with a little effort the CSS changes in this PR can be isolated to affect only variable dropdowns.

Current Behavior before this PR:
----
- In Action, Dropdown "closed"
<img width="547" height="147" alt="image" src="https://github.com/user-attachments/assets/360994a3-7e9c-412d-84c7-86dd44b13e9c" />

----

- In Action Dropdown, open

<img width="510" height="246" alt="image" src="https://github.com/user-attachments/assets/63862b3c-5fda-47d2-94e2-cc56b6d267aa" />

----
Unchanged:
| In Button Text  (v4.1, v4.2 text-mode)| v4.2 expression-mode|
|---|---|
|<img width="419" height="304" alt="image" src="https://github.com/user-attachments/assets/46e10d47-e078-4f3d-9107-12be20cdabdb" /> |  <img width="557" height="115" alt="image" src="https://github.com/user-attachments/assets/e6d0e6b2-5c97-400e-ae6f-3631b2662152" /> |

arikorn commented:
The final commit, above, formats "options" the same as in the button-text suggestions and also formats the "singe-value" similarly, but there may be better/other ways to do the same... maybe it needs a Custom SelectContainer?  

Incidentally, there's a difference in what the single-value box shows when clicked between  actions (set custom variable -- shows same text either way) and feedbacks (variable check value -- shows only the value when selected or has focus as shown below) -- I think because the latter allows the user to type in the value.

<img width="477" height="146" alt="image" src="https://github.com/user-attachments/assets/01cf1f29-8b1c-4de1-ae02-329e57e13fcf" />

<BR/>
<BR/>
Open:
<BR/>
<img width="473" height="316" alt="image" src="https://github.com/user-attachments/assets/ed87d0df-73cb-4107-855c-377fb492b21b" />

Julusian commented on webui/src/scss/_controls.scss:
```diff
@@ -152,6 +167,19 @@ label.disabled {
+/* Allow \n line-breaks and remove "excess" line spacing between pulldown options. */
+.select-control__option {
+	white-space: pre-wrap !important;
+	line-height: 1.1;
```
These feel a bit squished together to me, but that could just be because I am used to how the variables one looks

Julusian commented on webui/src/Controls/InternalModuleField.tsx:
```diff
@@ -225,6 +227,28 @@ export const InternalPageIdDropdown = observer(function InternalPageDropdown({
 })
 
+// Formatting for variable pulldown options:
+const CustomOption = React.memo((props: OptionProps<DropdownChoiceInt>) => {
```
As these are similar (identical?) to the ones for the text input dropdown, perhaps it should be the same component in a 'DropdownUtil.tsx'?
There might be some differences though which makes this not viable.

arikorn replied:
Done. Combined as described in  my response to your comment on line R28.

Julusian commented on webui/src/scss/_controls.scss:
```diff
@@ -37,19 +37,34 @@ label.disabled {
+	}
+}
+
+/* note that using `span` instead of `div` doesn't format correctly for the ellipsis */
```
This ellipsis is probably something we want for the other version of this too?

Julusian commented on webui/src/Controls/InternalModuleField.tsx:
```diff
@@ -225,6 +227,28 @@ export const InternalPageIdDropdown = observer(function InternalPageDropdown({
+})
+
+// Formatting for variable "single value" (shown when dropdown is closed) -- uses a different CSS class
+const CustomSingleValue = React.memo((props: OptionProps<DropdownChoiceInt>) => {
```
As with elsewhere, this feels rather squeezed. Visually it doesnt match how we do padding elsewhere.

I know there have been some comments that we have too much spacing/padding, but I think being consistent is more important

arikorn commented on webui/src/Components/DropdownInputField.tsx:
```diff
@@ -23,6 +23,9 @@ interface DropdownInputFieldProps {
 	checkValid?: (value: DropdownChoiceId) => boolean
+	// The following is a compromise since if we just use typeof components,
+	// Typescript objects to `OptionProps<DropdownChoiceInt>` in the custom property functions in InternalModuleField.tsx
+	customComponents?: Partial<Record<keyof typeof components, any>>
```
Done but it may be more than you anticipated, and quite honestly it doesn't really improve upon this particular `type`-ing mess.  Yet... maybe what I did is a good thing in itself anyway. Also, considering your comment on line R231, I took a slightly different approach:

The the current "custom" code and formatting appeared to be specific to variables since it uses `DropdownChoiceInt` which is defined in a misleadingly-named file, _LocalVariableDefinitions.tsx_. In fact it's used for all types of variables and even some things that aren't variables. So I did a bit of refactoring and renamed the file _DropDownInputFancy.tsx_, which now includes the `CustomOption` (which is shared with text input) & `ConstomSingle` functions too. So now any caller can request fancy-formatted two-line input as long as the options conform to  `DropdownChoiceInt`.)

arikorn commented:
This may or may not help... here is a visual aid to the effect line-height on the drop-down menu, "actual size":

<img width="453" height="230" alt="image" src="https://github.com/user-attachments/assets/2a02e036-3fc7-4d72-a152-7141e1769ca2" />

<br/><br/>

And here's a mapping of  line-height to `suggestLineHeight` for the monaco editor

|Line-height | Row Height (pix) |
|---|---|
| 1.5 | 59 |
| 1.4 | 56 |
| 1.3 | 53 |
| 1.2 | 51 |
| 1.1 | 48 |
| 1.0 | 45 |


These were done by adjusting line-height in the CSS for `.select-control__option, .variable-select-root__option` in _controls.scss (which affects all dropdown menus but monaco) and measuring the height of the blue section. For the monaco editor suggestions, `line-height` in `.main` (_expressions.scss_) should be set to around 40% of  `suggestLineHeight`.

Julusian commented:
monaco being different didnt bother me. Perhaps it was partly laziness (I avoided touching the styling as much as possible, and partly because I didnt want to spent days going down the rabbit hole of trying to make it look good/match.

I am not keen on the spacing change though:
<img width="608" height="286" alt="image" src="https://github.com/user-attachments/assets/5560de15-5e96-4420-95c6-dec9e32310a9" />
partly because its visibly uneven, with no padding above the first line, but loads below the second line.

I am also not 100% sold on the font size of those suggestions, purely because I think the suggestions are now larger than the actual text field.
----

For me the main thing is that the rest of the dropdowns should be visually consistent with each other and the rest of the app. So I dont think that reducing spacing within the dropdowns is a good thing to do right now, because while they can all match, they wont match the rest of companion.  
(yeah I suppose I am arguing for styling monaco, but I really dont enjoy fighting css..)

----

My only other comment is that I dont like the size change of this based upon where the focus is

https://github.com/user-attachments/assets/f4009809-4762-4b69-84d2-142def3ad07a

arikorn commented:
I fixed your three issues, see details below. That said, if you're still not satisfied with the monaco suggest-widget changes, lets just revert it for now: the main goal of this PR is to make the "standard" dropdowns more readable/neatly formatted. Monaco was really just a "bonus" change since the text is already formatted in two lines, so it's just a bit of icing...

To summarize this PR: 
- This PR formats action & feedback dropdown menus that select Variables to match the variable suggestions in button text (not button expression text), i.e. two lines, the first line boldface name of the variable. 
- A side-effect/bonus is that it allows developers to make two-line entries using `'\n'` in any dropdown menu.
- A second side-effect is allowing menu row-heights to be adjusted in one place (_control.scss_) for all callers of `DropdownInputField` and `TextInputField`. Previously line-height was effectively 1.5; I've set it to 1.3 in the latest commit for both the "closed-menu" value and the "open-menu" options (see my previous comment for size comparisons). While it's easy enough to set to 1.5, that big seemed "too-much" on the closed-menu values, and I suspect only the extreme "connoisseur" would notice a difference vs. static parts of the UI. (And as I pointed out earlier, the buttons themselves use 1.1 line-heights...). And just to reiterate -- please feel free to change this or request me to do so if you still disagree. 
- As a bonus in this PR, I adjusted the monaco suggest-widget to look more like the others. This change is _not_ essential to the PR and can be broken off as a separate PR (or abandoned) if it's not working out.

Details
---

1. Input box size change: **_Fixed_**. Notes:

    > I [Julian] dont like the size change of this based upon where the focus is

   Size change based on focus is a side-effect of the fact that when `allowCustom = true`, and the user selects the input, it bypasses `CustomSingleValue`.  I couldn't find a direct fix, so I added a special CSS class to fix it.  (Personally, I would have just set `allowCustom` to false, but a comment on the relevant line makes it clear that someone feels it is important. No problem!)

   Size is now stable, though text jumps down a bit to stay centered.

2. Vertical centering of options (rows) in monaco suggest widget: **_Fixed_**.
3. Width of the monaco suggest widget: **_Improved_**. Notes:

    This is the pre-existing behavior and is independent of font size (see examples below, using the latest beta version in `main`). I couldn't find a fully satisfactory dynamic-sizing solution because monaco didn't let me wrap the text (or should I say, I couldn't figure out the magic "key" to this door), so with dynamic sizing, when the parent-window is too narrow, the text is clipped too much. 

    My current best solution is to make it "reasonably wide" but not too wide.  I added `width: 25em` under `.monaco-editor .suggest-widget`, which seems the ideal size given the current font size.

Monaco `suggest-widget` behavior prior to this PR
---
Comparison of this PR (top-half) to main (below red line)
<img width="676" height="302" alt="image" src="https://github.com/user-attachments/assets/9dbd816e-10cf-48b3-8457-d902a9352053" />

<br/><br/>
Another example of the suggest-widget width in main (in an action expression)

<img width="512" height="195" alt="image" src="https://github.com/user-attachments/assets/6db1f061-7381-45f3-938b-fc3545a76e4a" />
